### PR TITLE
 Fix coverity issue in wurl_http_request function

### LINKED ISSUE #9169: Coverity issue in wurl_http_request function
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| master (4.3) - 5.0-dev | wazuh-modulesd (Github and Office 365 module) | Manager/Agent | Any | Any |

Hi team!

The last coverity report shows the next new defect related to `wurl_http_request` (shared/url.c)
```
New defect(s) Reported-by: Coverity Scan
Showing 1 of 1 defect(s)


** CID 236101:    (CHECKED_RETURN)
/shared/url.c: 470 in wurl_http_get_with_header()
/shared/url.c: 471 in wurl_http_get_with_header()
/shared/url.c: 466 in wurl_http_get_with_header()
/shared/url.c: 469 in wurl_http_get_with_header()
/shared/url.c: 467 in wurl_http_get_with_header()
/shared/url.c: 439 in wurl_http_get_with_header()
/shared/url.c: 434 in wurl_http_get_with_header()
/shared/url.c: 468 in wurl_http_get_with_header()


________________________________________________________________________________________________________
*** CID 236101:    (CHECKED_RETURN)
/shared/url.c: 470 in wurl_http_get_with_header()
464         req_header.size = 0;    /* no data at this point */
465     
466         curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
467         curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&req);
468         curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
469         curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, WriteMemoryCallback);
>>>     CID 236101:    (CHECKED_RETURN)
>>>     Calling "curl_easy_setopt(curl, _curl_opt, (void *)&req_header)" without checking return value. This library function may fail and return an error code.
470         curl_easy_setopt(curl, CURLOPT_HEADERDATA, (void *)&req_header);
471         curl_easy_setopt(curl, CURLOPT_URL, url);
472     
473         res = curl_easy_perform(curl);
474     
475         if (res != CURLE_OK) {
/shared/url.c: 471 in wurl_http_get_with_header()
465     
466         curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
467         curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&req);
468         curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
469         curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, WriteMemoryCallback);
470         curl_easy_setopt(curl, CURLOPT_HEADERDATA, (void *)&req_header);
>>>     CID 236101:    (CHECKED_RETURN)
>>>     Calling "curl_easy_setopt(curl, _curl_opt, url)" without checking return value. This library function may fail and return an error code.
471         curl_easy_setopt(curl, CURLOPT_URL, url);
472     
473         res = curl_easy_perform(curl);
474     
475         if (res != CURLE_OK) {
476             mdebug1("curl_easy_perform() failed: %s", curl_easy_strerror(res));
/shared/url.c: 466 in wurl_http_get_with_header()
460         req.memory = malloc(1);  /* will be grown as needed by the realloc above */
461         req.size = 0;    /* no data at this point */
462     
463         req_header.memory = malloc(1);  /* will be grown as needed by the realloc above */
464         req_header.size = 0;    /* no data at this point */
465     
>>>     CID 236101:    (CHECKED_RETURN)
>>>     Calling "curl_easy_setopt(curl, _curl_opt, WriteMemoryCallback)" without checking return value. This library function may fail and return an error code.
466         curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
467         curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&req);
468         curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
469         curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, WriteMemoryCallback);
470         curl_easy_setopt(curl, CURLOPT_HEADERDATA, (void *)&req_header);
471         curl_easy_setopt(curl, CURLOPT_URL, url);
/shared/url.c: 469 in wurl_http_get_with_header()
463         req_header.memory = malloc(1);  /* will be grown as needed by the realloc above */
464         req_header.size = 0;    /* no data at this point */
465     
466         curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
467         curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&req);
468         curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
>>>     CID 236101:    (CHECKED_RETURN)
>>>     Calling "curl_easy_setopt(curl, _curl_opt, WriteMemoryCallback)" without checking return value. This library function may fail and return an error code.
469         curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, WriteMemoryCallback);
470         curl_easy_setopt(curl, CURLOPT_HEADERDATA, (void *)&req_header);
471         curl_easy_setopt(curl, CURLOPT_URL, url);
472     
473         res = curl_easy_perform(curl);
474     
/shared/url.c: 467 in wurl_http_get_with_header()
461         req.size = 0;    /* no data at this point */
462     
463         req_header.memory = malloc(1);  /* will be grown as needed by the realloc above */
464         req_header.size = 0;    /* no data at this point */
465     
466         curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
>>>     CID 236101:    (CHECKED_RETURN)
>>>     Calling "curl_easy_setopt(curl, _curl_opt, (void *)&req)" without checking return value. This library function may fail and return an error code.
467         curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&req);
468         curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
469         curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, WriteMemoryCallback);
470         curl_easy_setopt(curl, CURLOPT_HEADERDATA, (void *)&req_header);
471         curl_easy_setopt(curl, CURLOPT_URL, url);
472     
/shared/url.c: 439 in wurl_http_get_with_header()
433             if (NULL != cert) {
434                 curl_easy_setopt(curl, CURLOPT_CAINFO, cert);
435             }
436         }
437     #endif
438     
>>>     CID 236101:    (CHECKED_RETURN)
>>>     Calling "curl_easy_setopt(curl, _curl_opt, "GET")" without checking return value. This library function may fail and return an error code.
439         curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "GET");
440     
441         headers = curl_slist_append(headers, "User-Agent: curl/7.58.0");
442     
443         if (headers == NULL) {
444             curl_easy_cleanup(curl);
/shared/url.c: 434 in wurl_http_get_with_header()
428     #ifndef WIN32
429         char const *cert = find_cert_list();
430     
431         // Enable SSL check if url is HTTPS
432         if(!strncmp(url,"https",5)){
433             if (NULL != cert) {
>>>     CID 236101:    (CHECKED_RETURN)
>>>     Calling "curl_easy_setopt(curl, _curl_opt, cert)" without checking return value. This library function may fail and return an error code.
434                 curl_easy_setopt(curl, CURLOPT_CAINFO, cert);
435             }
436         }
437     #endif
438     
439         curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "GET");
/shared/url.c: 468 in wurl_http_get_with_header()
462     
463         req_header.memory = malloc(1);  /* will be grown as needed by the realloc above */
464         req_header.size = 0;    /* no data at this point */
465     
466         curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteMemoryCallback);
467         curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void *)&req);
>>>     CID 236101:    (CHECKED_RETURN)
>>>     Calling "curl_easy_setopt(curl, _curl_opt, headers)" without checking return value. This library function may fail and return an error code.
468         curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
469         curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, WriteMemoryCallback);
470         curl_easy_setopt(curl, CURLOPT_HEADERDATA, (void *)&req_header);
471         curl_easy_setopt(curl, CURLOPT_URL, url);
472     
473         res = curl_easy_perform(curl);
```

Regards,
Nico

### PR DESCRIPTION
**Pull request by Nicogp**
|Related issue|
|---|
|Closes #9169 |

## Description

This PR fix coverity issues in wurl_http_request function and add new UT to that function

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [x] MAC OS X
- [x] Source installation
- [x] Package installation
- [x] Source upgrade
- [x] Package upgrade

- Memory tests for Linux
  - [x] Coverity

### CODE REVIEW

File: src/shared/url.c [hunk 1]
```diff
@@ -488,15 +488,24 @@ curl_response* wurl_http_request(char *method, char **headers, const char* url,
...
+    res += curl_easy_setopt(curl, CURLOPT_HEADERFUNCTION, WriteMemoryCallback);
+    res += curl_easy_setopt(curl, CURLOPT_HEADERDATA, (void *)&req_header);
+    res += curl_easy_setopt(curl, CURLOPT_URL, (void *)url);
 
     if (payload) {
-        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, (void *)payload);
+        res += curl_easy_setopt(curl, CURLOPT_POSTFIELDS, (void *)payload);
+    }
+
+    if (res != 0) {
```
Comment (cborla) [inline src/shared/url.c:502]: could be CURLE_OK

**Review by cborla — DISMISSED** [review]
LGTM

**Comment by jnasselle** [comment]
Hi @Nicogp !

It's not related to this issue, but could you please add a newline at the end of `wm_control.h`? this is generating a lot of warnings in Legacy Linux building.

Thanks!
Nico

File: src/shared/url.c (hunk 1 above)
Comment (Nicogp) [reply src/shared/url.c]: Done!

**Comment by Nicogp** [comment]
Hi @jnasselle!
Yes, of course! I added it in the last commit.
Thanks!

**Review by TomasTurina — APPROVED** [review]
LGTM!

**Pull request merged**